xonsh/main.py

```diff
@@ -253,6 +254,33 @@ def premain(argv=None):
 
 
 def main(argv=None):
+    try:
+        return main_xonsh(argv)
+    except Exception:
+        traceback.print_exc()
+        print('Xonsh encountered an issue during launch', file=sys.stderr)
+        foreign_shell = None
+        shells_file = '/etc/shells'
+        if not os.path.exists(shells_file):
+            return
+        excluded_list = ['xonsh', 'screen']
+        with open(shells_file) as f:
+            for line in f:
+                if not line.strip() or line.strip().startswith('#'):
+                    continue
+                if '/' not in line:
+                    continue
+                _, shell = line.strip().rsplit('/', 1)
+                if shell in excluded_list:
+                    continue
+                foreign_shell = line.strip()
+                break
+        if foreign_shell:
+            print('Failback to {}'.format(foreign_shell), file=sys.stderr)
+            os.execlp(foreign_shell, foreign_shell)
```

Comment:
It would be good if the exception body here were broken out into its own function.  try-excepts should only be a couple of lines long, ideally.